### PR TITLE
Fix joining of f-strings with different quotes when using quote style `Preserve`

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

Fixes https://github.com/astral-sh/ruff/issues/15514

Fixes joining implicit concatenated strings when using `quote-style="Preserve` if the parts use different quotes and at least one contains a nested string using quotes. 

The bug rooted from always returning `Preserve` in the `choose_quotes` function when using `quote-style="Preserve"` even in cases where f-strings are nested. 
This is fine for regular f-strings (because we can assume that the quotes were valid in the source) but it isn't sufficient when joining f-strings. 

## Test Plan

Added tests

### CODE REVIEW

**Comment by MichaReiser** [comment]
I've to run now but I want to spend some more time on adding more tests. @dhruvmanila maybe you have some ideas of other edge cases worth testing. But I think this is otherwise ready to go.

File: crates/ruff_python_formatter/resources/test/fixtures/ruff/expression/join_implicit_concatenated_string_preserve.py [hunk 1]
```diff
@@ -11,3 +11,12 @@
...
 f"{'Hy "User"'}" f'{"Hy 'User'"}'
+
+
+# Regression tests for https://github.com/astral-sh/ruff/issues/15514
+params = {}
+string = "this is my string with " f'"{params.get("mine")}"'
+string = f'"{params.get("mine")} ' f"with {'nested single quoted string'}"
+string = f"{'''inner ' '''}" f'{"""inner " """}'
+string = f"{10 + len('bar')=}" f"{10 + len('bar')=}"
+string = f"{10 + len('bar')=}" f'{10 + len("bar")=}'
```
Comment (dhruvmanila) [inline crates/ruff_python_formatter/resources/test/fixtures/ruff/expression/join_implicit_concatenated_string_preserve.py:22]: We might want to add some test cases which includes format spec.

Is the following correct?
```diff
 params = {}
-string = "this is my string with " f'"{params.get("mine")=:{"xxx"}}"'
+string = f'this is my string with "{params.get("mine")=:{"xxx"}}"'
```

with quote-style `preserve` and < 3.12 Python.

Or, is there no way to preserve the quote style here?

Comment (dhruvmanila) [reply crates/ruff_python_formatter/resources/test/fixtures/ruff/expression/join_implicit_concatenated_string_preserve.py:22]: This doesn't seem to be relevant to this PR though.

**Review by dhruvmanila — APPROVED** [review]

File: crates/ruff_python_formatter/resources/test/fixtures/ruff/expression/join_implicit_concatenated_string_preserve.py (hunk 1 above)
Comment (MichaReiser) [reply crates/ruff_python_formatter/resources/test/fixtures/ruff/expression/join_implicit_concatenated_string_preserve.py:22]: There's no strictly correct answer when it comes to preserve and joining implicit concatenated strings if not all parts use the same quotes. You have to choose which quotes you want to preserve and in this example, the formatter decides to preserve the quotes from the second part because the string can't be joined otherwise.

Comment (dhruvmanila) [reply crates/ruff_python_formatter/resources/test/fixtures/ruff/expression/join_implicit_concatenated_string_preserve.py:22]: That sounds reasonable, thanks
